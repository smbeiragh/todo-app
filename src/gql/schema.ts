import { gql } from 'apollo-server-express';
import {FakeDB} from "./mockDB";

export const typeDefs = gql`
  type ToDo {
    id: String
    title: String
    priority: Int
    dueDate: String
    done: Boolean
  }
  
  type Query {
    getToDos(orderBy: String, orderType: String): [ToDo]
  }
  
  type Mutation {
    createToDo(title: String!, dueDate: String, priority: Int, orderBy: String, orderType: String): [ToDo]
    deleteToDo(id: String!): String
    updateToDo(id: String!, title: String, dueDate: String, priority: Int, done: Boolean): ToDo
  }
`;

export const createResolvers = (fakeDB: FakeDB) => ({
  Query: {
    getToDos: (root: any, {orderBy = 'title', orderType = 'ASC'} = {}) => {
      return fakeDB.getToDos({orderBy, orderType});
    }
  },
  Mutation: {
    createToDo: (
        root: any,
        { title = '', priority = 1, dueDate, done = false ,orderBy = 'title', orderType= 'ASC'}
        : { title: string, priority: number, dueDate: string, done: boolean, orderBy: string, orderType: string}
      ) => {
      fakeDB.createToDo({
        title,
        priority,
        dueDate,
        done
      });
      return fakeDB.getToDos({orderBy, orderType});
    },
    deleteToDo: (root: any, { id = '' } = {}) => {
      return fakeDB.deleteToDo({id});
    },
    updateToDo: (root: any, { id ,title, priority, dueDate, done} : ITodo = ({} as any)) : ITodo => {
      return fakeDB.updateToDo({ id ,title, priority, dueDate, done});
    },
  }
});
