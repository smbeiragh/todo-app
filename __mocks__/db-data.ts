export function creatData(): Array<ITodoInput> {
  const toDos: Array<ITodoInput> = [
    {
      title: 'Todo 1',
      priority: 2,
      dueDate: '2018-09-27T10:53:48.248Z',
      done: false
    },
    {
      title: 'Todo 2',
      priority: 1,
      dueDate: '2018-09-27T10:53:48.248Z',
      done: false
    }
  ];
  return toDos;
}
