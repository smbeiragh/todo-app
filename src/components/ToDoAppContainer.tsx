import * as React from 'react';
import { connect } from 'react-redux';
import { sortAction } from './../redux/actions';
import { ToDoApp } from './ToDoApp';

const mapStateToProps = (state: any) => {
  return {
    orderBy: state.todos.orderBy
  };
};

const mapDispatchToProps = (dispatch: any) => {
  return {
    onSort: (orderBy: string) => {
      dispatch(sortAction({ orderBy }));
    }
  };
};

export const ToDoAppContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(ToDoApp);