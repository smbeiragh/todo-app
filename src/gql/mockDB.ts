const defaultDate = () :string => (new Date((new Date()).setDate((new Date()).getDate()+7))).toISOString();

class ToDo {
  id: string;
  title: string;
  priority: number;
  dueDate: string;
  dueDateTimeStamp: number;
  done: boolean;

  constructor({ title = '', priority = 1, dueDate = defaultDate() , done = false }: ITodoInput) {
    this.id = require('crypto').randomBytes(10).toString('hex'),
      this.title = title,
      this.priority = priority,
      this.dueDate = dueDate,
      this.dueDateTimeStamp = +(new Date(dueDate));
    this.done = done
  }

  setDueDate(dueDate: string) {
    this.dueDate = dueDate;
    this.dueDateTimeStamp = +(new Date(dueDate));
    return this;
  }
}

function sort(fakeDB: Array<ToDo> ,{orderBy = 'title', orderType = 'ASC'} = {}) {
  const fieldName = orderBy === 'dueDate' ? 'dueDateTimeStamp' : orderBy;
  if (orderType === 'ASC') {
    return fakeDB.sort((a, b) => (a as any)[fieldName] < (b as any)[fieldName] ? -1 : 1 )
  }
  return fakeDB.sort((a, b) => (a as any)[fieldName] > (b as any)[fieldName] ? -1 : 1)
}

function defaultInitData() {
  const data: Array<ITodoInput> = [
    {
      title: 'Create a Readme',
      priority: 2,
      dueDate: (new Date()).toISOString(),
      done: false
    },
    {
      title: 'Finish The project',
      priority: 1,
      dueDate: (new Date()).toISOString(),
      done: false
    },
    {
      title: 'Host the project on github',
      priority: 1,
      dueDate: (new Date()).toISOString(),
      done: false
    }
  ];

  return data;
}

export class FakeDB {

  data: Array<ToDo>;

  constructor(initData: Array<ITodoInput> = defaultInitData()) {
    this.data = initData.map((item) :ToDo => { return new ToDo(item) });
  }

  getToDos({orderBy = 'title', orderType = 'ASC'} = {}) {
    return sort(this.data, {orderBy, orderType});
  }

  createToDo({ title = '', priority = 1, dueDate, done = false }: ITodoInput) {
    const todo = new ToDo({
      title,
      priority,
      dueDate,
      done
    });
    this.data.push(todo);
    return todo;
  }

  deleteToDo({ id = '' } = {}) {
    const index =  (this.data as any).findIndex((item: any)=> id === item.id);
    if(index > -1) {
      this.data.splice(index, 1);
      return id;
    }
    return null;
  }

  updateToDo({ id ,title, priority, dueDate, done}: ITodo) : ITodo {
    if(id) {
      const todo =  (this.data as any).find((item: any)=> id === item.id);
      if (!todo) {
        return null;
      }
      if (title) {
        todo.title = title;
      }
      if (typeof priority === 'number') {
        todo.priority = priority;
      }
      if (dueDate) {
        todo.setDueDate(dueDate);
      }
      if (typeof done === "boolean") {
        todo.done = done;
      }

      return todo;
    }
    return null;
  }
}