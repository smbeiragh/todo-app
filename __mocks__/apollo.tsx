import * as path from 'path';
import { Server } from 'http';
import express from 'express';
import { createResolvers, typeDefs } from "../src/gql/schema";
import { ApolloServer } from "apollo-server-express";
import { ApolloProvider, renderToStringWithData } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { createHttpLink } from 'apollo-link-http';
import { InMemoryCache } from "apollo-cache-inmemory";
import * as React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import fetch from 'node-fetch';
import { rootReducer } from '../src/redux/reducers/index';
import { testAction } from '../src/redux/actions/index';
import { FakeDB } from '../src/gql/mockDB';

export function createServer(fakeDB: FakeDB) {
  const app = express();
  const server = new Server(app);

  const gqlServer = new ApolloServer({
    // These will be defined for both new or existing servers
    typeDefs,
    resolvers: createResolvers(fakeDB),
  });

  gqlServer.applyMiddleware({ app });

  // start the server
  const port = process.env.NODE_PORT;

  return new Promise((resolve: any, reject: any) => {
    server.listen(port, (err: any) => {
      if (err) {
        console.error(err);
        reject(err);
      } else {
        resolve(server);
        console.log(process.env.NODE_ENV);
        console.info(`Server is running on http://localhost:${port}`);
      }
    });
  });

}

export function createClient() {
  const store = createStore(rootReducer);

  // start the server
  const port = process.env.NODE_PORT;

  // lets put something in redux store
  store.dispatch(testAction({text: 'some test string!'}));

  const initState = JSON.stringify(store.getState());

  const apolloClient = new ApolloClient({
    ssrMode: true,
    // Remember that this is the interface the SSR server will use to connect to the
    // API server, so we need to ensure it isn't firewalled, etc
    link: createHttpLink({
      uri: `http://127.0.0.1:${port}/graphql`,
      // credentials: 'same-origin',
      fetch : fetch as any
    }),
    cache: new InMemoryCache(),
  });

  return { store, apolloClient };
}


