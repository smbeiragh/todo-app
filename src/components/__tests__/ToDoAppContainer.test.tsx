import * as React from 'react';
import { ToDoAppContainer } from '../ToDoAppContainer';
import * as renderer from 'react-test-renderer';
import { shallow, mount, render } from 'enzyme';
import { ApolloProvider, getDataFromTree } from 'react-apollo';
import { Provider } from 'react-redux';
import { createServer, createClient } from '../../../__mocks__/apollo';
import {FakeDB} from './../../gql/mockDB';
import { creatData } from '../../../__mocks__/db-data';

let store: any;
let apolloClient: any;
let server: any;

beforeAll(() => {
  return createServer(new FakeDB(creatData()))
    .then((_server: any) => {
      server = _server;
      return new Promise((res: any, rej: any) => {
        setTimeout(() => {
          const all = createClient();
          store = all.store;
          apolloClient = all.apolloClient;
          res(server);
        }, 100);
      })
    });
});

afterAll((done) => {
  if (server) {
    server.close(done);
  } else {
    done()
  }
});

test('ToDoAppContainer should render', () => {

  const App = (
    <ApolloProvider client={apolloClient}>
      <Provider store={store}>
        <ToDoAppContainer />
      </Provider>
    </ApolloProvider>
  );

  return getDataFromTree(App).then(() => {
    const component = renderer.create(App);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});
