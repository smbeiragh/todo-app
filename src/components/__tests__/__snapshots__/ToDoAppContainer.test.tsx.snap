// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoAppContainer should render 1`] = `
<div
  className="c-toDoApp"
>
  <h1
    className="c-toDoApp__header"
  >
    Todo list
  </h1>
  <div
    className="c-toDoApp__form c-toDoAddForm"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        className="c-toDoAddForm__input"
        maxLength={300}
        onChange={[Function]}
        placeholder="Enter Todo Name and press Enter"
        type="text"
        value=""
      />
    </form>
  </div>
  <div
    className="c-toDoSort"
  >
    <div
      className="c-toDoSort__label"
    >
      Sort By
    </div>
    <button
      className="c-toDoSort__btn"
      onClick={[Function]}
    >
      Name
    </button>
    <button
      className="c-toDoSort__btn"
      onClick={[Function]}
    >
      Due Date
    </button>
    <button
      className="c-toDoSort__btn"
      onClick={[Function]}
    >
      Priority
    </button>
  </div>
  <ul
    className="c-toDoList"
  >
    <li
      className="c-toDoList__item c-toDo"
    >
      <div
        className="c-toDo__done"
      >
        <label
          className="c-coolCheckBox"
        >
          <input
            checked={false}
            className="c-coolCheckBox__input"
            onChange={[Function]}
            type="checkbox"
          />
          <i
            className="c-coolCheckBox__check icon-done"
          />
        </label>
      </div>
      <div
        className="c-toDo__title"
      >
        Todo 2
      </div>
      <div
        className="c-toDo__dueDate c-toDoDueDatePicker"
      >
        <div>
          <div
            className="react-datepicker-wrapper"
          >
            <div
              className="react-datepicker__input-container"
            >
              <input
                className="c-toDoDueDatePicker__input"
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={false}
                type="text"
                value="09/27/2018"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="c-toDo__priority c-priorityInput"
      >
        <button
          className="c-priorityInput__btn c-priorityInput__btn--up"
          disabled={false}
          onClick={[Function]}
        >
          <i
            className="c-priorityInput__icon icon-less"
          />
        </button>
        <div
          className="c-priorityInput__display"
        >
          medium
        </div>
        <button
          className="c-priorityInput__btn c-priorityInput__btn--down"
          disabled={false}
          onClick={[Function]}
        >
          <i
            className="c-priorityInput__icon icon-more"
          />
        </button>
      </div>
      <div
        className="c-toDo__remove"
      >
        <button
          className="c-toDo__removeBtn"
          onClick={[Function]}
        >
          <i
            className="c-toDo__removeIcon icon-close"
          />
        </button>
      </div>
    </li>
    <li
      className="c-toDoList__item c-toDo"
    >
      <div
        className="c-toDo__done"
      >
        <label
          className="c-coolCheckBox"
        >
          <input
            checked={false}
            className="c-coolCheckBox__input"
            onChange={[Function]}
            type="checkbox"
          />
          <i
            className="c-coolCheckBox__check icon-done"
          />
        </label>
      </div>
      <div
        className="c-toDo__title"
      >
        Todo 1
      </div>
      <div
        className="c-toDo__dueDate c-toDoDueDatePicker"
      >
        <div>
          <div
            className="react-datepicker-wrapper"
          >
            <div
              className="react-datepicker__input-container"
            >
              <input
                className="c-toDoDueDatePicker__input"
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={false}
                type="text"
                value="09/27/2018"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="c-toDo__priority c-priorityInput"
      >
        <button
          className="c-priorityInput__btn c-priorityInput__btn--up"
          disabled={true}
          onClick={[Function]}
        >
          <i
            className="c-priorityInput__icon icon-less"
          />
        </button>
        <div
          className="c-priorityInput__display"
        >
          high
        </div>
        <button
          className="c-priorityInput__btn c-priorityInput__btn--down"
          disabled={false}
          onClick={[Function]}
        >
          <i
            className="c-priorityInput__icon icon-more"
          />
        </button>
      </div>
      <div
        className="c-toDo__remove"
      >
        <button
          className="c-toDo__removeBtn"
          onClick={[Function]}
        >
          <i
            className="c-toDo__removeIcon icon-close"
          />
        </button>
      </div>
    </li>
  </ul>
</div>
`;
